sql/stripe/funnel/view.sql

```diff
@@ -0,0 +1,96 @@
+CREATE OR REPLACE VIEW
+  funnel
+AS
+WITH subscriptions AS (
+  SELECT
+    plan,
+    status = "incomplete_expired" AS incomplete_expired,
+    cancel_at_period_end
+    OR "cancelled_for_customer_at" IN (SELECT key FROM UNNEST(metadata)) AS cancelled_for_customer,
+    -- count subscriptions from the day after they start to the day after they end
+    -- TODO: decide whether to count users at the end of the day or the beginning of the day
+    DATE_ADD(DATE(start_date), INTERVAL 1 DAY) AS start_date,
+    -- use day after ended_at to count cancelled and renew failed
+    DATE_ADD(DATE(ended_at), INTERVAL 1 DAY) AS end_date,
+    trial_end,
+    MIN(DATE_ADD(DATE(start_date), INTERVAL 1 DAY)) OVER (
+      PARTITION BY
+        customer
+    ) AS customer_start_date,
+    customer,
+    -- TODO: join on guardian logs and cloudSQL for mozilla vpn funnel information (attribution,
+    -- client running, vpn active, waitlist)
+  FROM
+    stripe_derived.subscriptions_v1
+  WHERE
+    status != "incomplete"
+),
+counts AS (
+  SELECT
+    `date`,
+    plan,
+    CASE
+    WHEN
+      incomplete_expired
+    THEN
+      "failed setup"
+    WHEN
+      TIMESTAMP(`date`) < trial_end
+    THEN
+      "trialing"
+    -- TODO decide whether subscriptions that go through a trial are ever counted as "new"
+    -- WHEN `date` = DATE_ADD(DATE(trial_end), INTERVAL 1 DAY) THEN "new"
+    WHEN
+      `date` = customer_start_date
+    THEN
+      "new"
+    WHEN
+      `date` = start_date
+    THEN
+      "resurrected"
+    WHEN
+      `date` = end_date
+      AND cancelled_for_customer
+    THEN
+      "cancelled"
+    WHEN
+      `date` = end_date
+    THEN
+      "failed renewal"
+    ELSE
+      "active"
+    END
+    AS status,
+    -- TODO decide whether to count customers or subscriptions, because in a single day a customer
```

Comment:
Yeah, I smell trouble here. We need to count the number of active (and trialling) subscriptions at UTC midnight each day, and we need to account for the day-to-day changes in those numbers (i.e. count new subscriptions, failed renewals, etc).

If one subscription can end in multiple ways (e.g. a payment fails then it's manually ended, perhaps on a later date) then we don't want to double-count its end (unless there's a balancing number of things counted in the "start" data).

Are we at risk of that here or does Stripe's data structure save us from it?

Reply:
W.r.t. "failed setup", I think that this kind of thing should be part of the funnel analysis; the visualisations we're currently aiming at should start at the point that the user has already signed up (or at least set up a trial).